stdlib/LinearAlgebra/src/uniformscaling.jl

```diff
@@ -400,3 +400,12 @@ Array(s::UniformScaling, dims::Dims{2}) = Matrix(s, dims)
 ## Diagonal construction from UniformScaling
 Diagonal{T}(s::UniformScaling, m::Integer) where {T} = Diagonal{T}(fill(T(s.λ), m))
 Diagonal(s::UniformScaling, m::Integer) = Diagonal{eltype(s)}(s, m)
+
+function dot(x::AbstractVector, J::UniformScaling, y::AbstractVector)
+    return mapreduce(t -> dot(t[1], J.λ, t[2]), +, zip(x, y))
+end
+function dot(x::AbstractVector, a::Number, y::AbstractVector)
```

Comment:
Since `J.λ <: Number` always, it is probably better to define
```jl
dot(x::AbstractVector, J::UniformScaling, y::AbstractVector) = dot(x, J.λ, y)
```
and then define
```jl
dot(x::AbstractVector, a::Number, y::AbstractVector) =
    mapreduce(t -> dot(t[1], a, t[2]), +, zip(x, y))
dot(x::AbstractVector, a::Union{Real,Complex}, y::AbstractVector) = a*dot(x, y)
```
so that we have a single implementation of the scalar case.